othello/game_logic/game.py

```diff
@@ -1,111 +1,119 @@
 
 from game_logic.agents.agent import Agent
 from game_logic.agents.trainable_agent import TrainableAgent
 from game_logic.board import Board
 from utils.color import Color
+from utils.config import Config
+from utils.global_config import GlobalConfig
 
 
 class Game:
-	def __init__(self, episode: int, black: Agent, white: Agent, board_size: int, verbose: bool, tournament_mode: bool = False):
-		# check arguments
-		assert black.color is Color.BLACK, f'Invalid black agent'
-		assert white.color is Color.WHITE, f'Invalid white agent'
-
+	def __init__(self, global_config: GlobalConfig, config: Config, episode: int):
+		self.global_config: GlobalConfig = global_config
+		self.config: Config = config
 		self.episode: int = episode
-		self.black: Agent = black
-		self.white: Agent = white
-		self.board: Board = Board(board_size) if not tournament_mode else Board(board_size, change_board_after_n_plays=1) # create a new board
-		self.verbose: bool = verbose
 
-		# state of the game
+		self.board: Board = Board(global_config.board_size) if not config.tournament_mode else Board(
+			global_config.board_size, change_board_after_n_plays=4)  # create board
 		self.ply = 0  # no plies so far
-		self.agent: Agent = black  # black begins
+		self.agent: Agent = self.config.black  # black always begins
 		self.prev_pass: bool = False  # opponent did not pass in previous ply
 		self.done: bool = False  # not done yet
 
 	def play(self) -> None:
-		if self.verbose:
-			print(f'Episode: {self.episode}')
-
-		if self.verbose:
-			print(f'Ply: {self.ply} (INIT)')
+		if self.config.verbose_live:
+			print(f'Episode {self.episode}:')
+			print(f'\tPly {self.ply}: INIT')
 			print(self.board)
 
 		# play until done
 		while not self.done:
 			# update
 			self.ply += 1
-			if self.verbose:
-				print(f'Ply: {self.ply} ({self.agent.color.name})')
+
+			if self.config.verbose_live:
+				disk_icon: str = u'\u25CF' if self.agent.color == Color.BLACK else u'\u25CB'
+				print(f'\tPly {self.ply}: {self.agent.color.name} {disk_icon}')
 
 			# get legal actions
 			legal_actions: dict = self.board.get_legal_actions(self.agent.color.value)
 
 			if not legal_actions:
 				# pass if no legal actions
-				if self.verbose:
-					print(f'No legal actions')
-					print(f'Next action: PASS')
+				if self.config.verbose_live:
+					print(f'\tNo legal actions')
+					print(self.board)
+					print(f'\tNext action: PASS')
 				if self.prev_pass:
 					self.done = True  # no agent has legal actions, deadlock
 				self.prev_pass = True  # this agent has no legal actions, pass
 			else:
 				# get next action from legal actions and take it
 				location, legal_directions = self.agent.get_next_action(self.board, legal_actions)
-				if self.verbose:
-					print(f'Legal actions: {list(legal_actions.keys())}')
-					print(f'Next action: {location}')
+				if self.config.verbose_live:
+					print(f'\tLegal actions: {list(legal_actions.keys())}')
+					board_copy: Board = copy.deepcopy(self.board)
```

Comment:
liever geen copy package gebruiken. Gebruik eerder self.board.deepcopyy()

Reply:
resolved in https://github.com/gartangh/Deep-Learning-Project/pull/18/commits/ddb8fd31f5155e71cdcc41aa60846491503e84e8